Admin DataGrid: Keep search terms after page reload

AnnikaWolff opened the pull request:
When searching in a DataGrid in the admin panel, the search term appears as a "post-it" beneath the search bar. 
![image](https://user-images.githubusercontent.com/57894757/84378347-5084e000-abe4-11ea-9adb-9b4d9c764dcf.png)

But as soon as a filter is added, e.g. a channel is selected, the search term disappears, although it remains as a URL parameter. The result is filtered correctly, but the user is confused, because the search term is not shown.

This PR fixes this issue: The search term will remain with all applied filters beneath the search bar. The channel and the customer group will appear there as well now.
That means, the gray terms are directly corresponding to the URL parameters.
![image](https://user-images.githubusercontent.com/57894757/84378979-58914f80-abe5-11ea-8fd0-df1ceed6df92.png)

EDIT: I just added encoding of the search terms, to let us be able to search for weird coupon codes as "AB#CD" or "All20%Off".

jitendra-webkul commented on packages/Webkul/Ui/src/DataGrid/DataGrid.php:
```diff
@@ -400,12 +400,12 @@ public function sortOrFilterCollection($collection, $parseInfo)
             } else {
                 foreach ($this->completeColumnDetails as $column) {
-                    if($column['index'] == $columnName && !$column['filterable']) {
+                    if($column['index'] === $columnName && !$column['filterable']) {
```
Add space after if

AnnikaWolff replied:
Done.

jitendra-webkul commented on packages/Webkul/Ui/src/Resources/views/datagrid/table.blade.php:
```diff
@@ -290,51 +290,52 @@
 
-                        for(colIndex in this.columns) {
+                        for(let colIndex in this.columns) {
                             if(this.columns[colIndex].index == this.columnOrAlias) {
```
Add space after if

AnnikaWolff replied:
Done.

jitendra-webkul commented on packages/Webkul/Ui/src/Resources/views/datagrid/table.blade.php:
```diff
@@ -290,51 +290,52 @@
                                 alert('{{__('ui::app.datagrid.zero-index')}}');
                             }
 
                             if(indexConditions)
```
Add space after if

AnnikaWolff replied:
Done

jitendra-webkul commented on packages/Webkul/Ui/src/Resources/views/datagrid/table.blade.php:
```diff
@@ -290,51 +290,52 @@
 
-                        for(colIndex in this.columns) {
-                            if(this.columns[colIndex].index == alias) {
+                        for(let colIndex in this.columns) {
```
Please add space after for

AnnikaWolff replied:
Done.

jitendra-webkul commented on packages/Webkul/Ui/src/Resources/views/datagrid/table.blade.php:
```diff
@@ -345,7 +346,7 @@
                         }
 
-                        for(id in this.massActions) {
+                        for(let id in this.massActions) {
```
Please add space after for

AnnikaWolff replied:
Done.

jitendra-webkul commented on packages/Webkul/Ui/src/Resources/views/datagrid/table.blade.php:
```diff
@@ -355,34 +356,34 @@
                     findCurrentSort: function() {
-                        for(i in this.filters) {
-                            if (this.filters[i].column == 'sort') {
+                        for(let i in this.filters) {
```
Please add space after for

jitendra-webkul commented on packages/Webkul/Ui/src/Resources/views/datagrid/table.blade.php:
```diff
@@ -355,34 +356,34 @@
-                            for(i in this.massActionTargets) {
-                                if (this.massActionTargets[i].type == 'delete') {
+                        if (this.massActionType === 'delete') {
+                            for(let i in this.massActionTargets) {
```
√

jitendra-webkul commented on packages/Webkul/Ui/src/Resources/views/datagrid/table.blade.php:
```diff
@@ -355,34 +356,34 @@
-                            for(i in this.massActionTargets) {
-                                if (this.massActionTargets[i].type == 'delete') {
+                        if (this.massActionType === 'delete') {
+                            for(let i in this.massActionTargets) {
```
Please add space after for

AnnikaWolff replied:
Done.

jitendra-webkul commented on packages/Webkul/Ui/src/Resources/views/datagrid/table.blade.php:
```diff
@@ -355,34 +356,34 @@
-                            for(i in this.massActionTargets) {
-                                if (this.massActionTargets[i].type == 'update') {
+                        if (this.massActionType === 'update') {
+                            for(let i in this.massActionTargets) {
```
Please add space after for

AnnikaWolff replied:
Done.

jitendra-webkul commented on packages/Webkul/Ui/src/Resources/views/datagrid/table.blade.php:
```diff
@@ -397,23 +398,26 @@
-                                        if (this.filters[j].column == column) {
-                                            if (this.filters[j].cond == condition && this.filters[j].val == response) {
-                                                filterRepeated = 1;
+                                    for(let j = 0; j < this.filters.length; j++) {
```
Please add space after for

AnnikaWolff replied:
Done. 

And found some more on my way through the file.
Thanks for reviewing :)

vaishaliwebkul commented:
Hi @AnnikaWolff 
Please correct this
![Screenshot_from_2020_06_18_12_56_26](https://user-images.githubusercontent.com/43210406/84991027-96512380-b163-11ea-9359-0dc3b06491ae.png)

AnnikaWolff commented:
Hi @vaishaliwebkul 

Yes, the search term is converted from a url encoded string to a url decoded string for the sake of better readability.
But I do not think, that is a problem, do you?

For instance, I have a simple product called "Simple Product". I search for it with the terms "Simple Product", having a space between the two words. In the URL, the search terms are encoded to "Simple%20Product", that is not nice to read. So coming out of the URL, the terms are decoded to "Simple+Product", which is easier to read. 

The search result is always the same: If looking for "Simple Product", "Simple%20Product" or "Simple+Product", I always find my simple product.

Concerning the failed check, I already notified @jitendra-webkul about the faulty migration causing those errors.

vaishaliwebkul commented:
@AnnikaWolff 

1. My concern is for search key changed when locale added; 
for example When I search for 'simple product' the search result be shown 'simple product', as and when I added the locale key 'en' then previous search key will be decoded to 'simple+product', I think it's a bug, we should keep the search key same before and after adding locale key.

2. Another Bug I just found while filtering list on the basis of status->True/Active (converted to False/Inactive)

![Screenshot (7)](https://user-images.githubusercontent.com/43210406/86342939-e953cc00-bc75-11ea-801d-6bb5e7ddf6ed.png)

AnnikaWolff commented:
Hi @vaishaliwebkul 

Alright, I will check that.

AnnikaWolff commented:
Dear @vaishaliwebkul 

1st: I removed the "+" symbol that appeared when searching for multiple terms and adding a locale filter.
2nd: I fixed the bug of the boolean filter always showing the "false / inactive" value.

And 3rd: I added a label to the locale filter.

![image](https://user-images.githubusercontent.com/57894757/87509329-97e20e80-c671-11ea-94ea-76dbd67f4d45.png)

Please feel free to test again.

vaishaliwebkul commented:
Hi @AnnikaWolff 
Thanks for resolving the above issue :)

One more update, In the cart rule page, filtered channels & customer groups tag with their ID, can't we show them by their name likewise in the product list.

![Screenshot (26)](https://user-images.githubusercontent.com/43210406/87689774-3d969a00-c7a6-11ea-8c43-1d8348f18064.png)


![Screenshot (26)](https://user-images.githubusercontent.com/43210406/87689539-f7413b00-c7a5-11ea-96c2-5f9b7240dfbf.png)

AnnikaWolff commented:
Hi @vaishaliwebkul 

This is much more effort because I have to refactor all DataGrids and their views at once. 
So stay tuned, it may take a while.

AnnikaWolff commented:
Hi  @vaishaliwebkul 

Done :) 
Filtered channels, locales and customer groups are now shown by their names instead of code or id.

And all my changes have another advantage, I suppose: It is much easier now to add those filters to other DataGrids, if needed.
